File: app/src/main/java/com/whatsapp/stickerimporter/StickerPack.java
package com.whatsapp.stickerimporter;

import android.os.Parcel;
import android.os.Parcelable;

import java.util.ArrayList;
import java.util.List;

/**
 * Represents a sticker pack containing multiple stickers
 */
public class StickerPack implements Parcelable {
    public String identifier;
    public String name;
    public String publisher;
    public String trayImageFile;
    public String publisherEmail;
    public String publisherWebsite;
    public String privacyPolicyWebsite;
    public String licenseAgreementWebsite;
    public String imageDataVersion;
    public boolean avoidCache;
    public long totalSize;
    public List<Sticker> stickers;
    public boolean isWhitelisted;

    public StickerPack(String identifier, String name, String publisher, String trayImageFile,
                       String publisherEmail, String publisherWebsite, String privacyPolicyWebsite,
                       String licenseAgreementWebsite) {
        this.identifier = identifier;
        this.name = name;
        this.publisher = publisher;
        this.trayImageFile = trayImageFile;
        this.publisherEmail = publisherEmail;
        this.publisherWebsite = publisherWebsite;
        this.privacyPolicyWebsite = privacyPolicyWebsite;
        this.licenseAgreementWebsite = licenseAgreementWebsite;
        this.stickers = new ArrayList<>();
        this.totalSize = 0;
    }

    protected StickerPack(Parcel in) {
        identifier = in.readString();
        name = in.readString();
        publisher = in.readString();
        trayImageFile = in.readString();
        publisherEmail = in.readString();
        publisherWebsite = in.readString();
        privacyPolicyWebsite = in.readString();
        licenseAgreementWebsite = in.readString();
        imageDataVersion = in.readString();
        avoidCache = in.readByte() != 0;
        totalSize = in.readLong();
        stickers = in.createTypedArrayList(Sticker.CREATOR);
        isWhitelisted = in.readByte() != 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeString(identifier);
        dest.writeString(name);
        dest.writeString(publisher);
        dest.writeString(trayImageFile);
        dest.writeString(publisherEmail);
        dest.writeString(publisherWebsite);
        dest.writeString(privacyPolicyWebsite);
        dest.writeString(licenseAgreementWebsite);
        dest.writeString(imageDataVersion);
        dest.writeByte((byte) (avoidCache ? 1 : 0));
        dest.writeLong(totalSize);
        dest.writeTypedList(stickers);
        dest.writeByte((byte) (isWhitelisted ? 1 : 0));
    }

    @Override
    public int describeContents() {
        return 0;
    }

    public static final Creator<StickerPack> CREATOR = new Creator<StickerPack>() {
        @Override
        public StickerPack createFromParcel(Parcel in) {
            return new StickerPack(in);
        }

        @Override
        public StickerPack[] newArray(int size) {
            return new StickerPack[size];
        }
    };

    public void setAndroidPlayStoreLink(String androidPlayStoreLink) {
        // Reserved for future use
    }

    public void setIosAppStoreLink(String iosAppStoreLink) {
        // Reserved for future use
    }
}

File: app/src/main/java/com/mystickers/whatsapp/StickerPack.kt
package com.mystickers.whatsapp

import android.os.Parcel
import android.os.Parcelable

/**
 * Represents a sticker pack that can be added to WhatsApp.
 * WhatsApp supports a maximum of 30 stickers per pack.
 */
data class StickerPack(
    val identifier: String,
    val name: String,
    val publisher: String,
    val trayImageFile: String,
    val publisherEmail: String = "",
    val publisherWebsite: String = "",
    val privacyPolicyWebsite: String = "",
    val licenseAgreementWebsite: String = "",
    val imageDataVersion: String = "1",
    val avoidCache: Boolean = false,
    val animatedStickerPack: Boolean = false,
    val stickers: MutableList<Sticker> = mutableListOf(),
    var isWhatsAppInstalled: Boolean = false
) : Parcelable {

    val totalSize: Long
        get() = stickers.sumOf { it.size }

    constructor(parcel: Parcel) : this(
        parcel.readString() ?: "",
        parcel.readString() ?: "",
        parcel.readString() ?: "",
        parcel.readString() ?: "",
        parcel.readString() ?: "",
        parcel.readString() ?: "",
        parcel.readString() ?: "",
        parcel.readString() ?: "",
        parcel.readString() ?: "1",
        parcel.readByte() != 0.toByte(),
        parcel.readByte() != 0.toByte(),
        mutableListOf<Sticker>().apply { parcel.readTypedList(this, Sticker.CREATOR) },
        parcel.readByte() != 0.toByte()
    )

    override fun writeToParcel(parcel: Parcel, flags: Int) {
        parcel.writeString(identifier)
        parcel.writeString(name)
        parcel.writeString(publisher)
        parcel.writeString(trayImageFile)
        parcel.writeString(publisherEmail)
        parcel.writeString(publisherWebsite)
        parcel.writeString(privacyPolicyWebsite)
        parcel.writeString(licenseAgreementWebsite)
        parcel.writeString(imageDataVersion)
        parcel.writeByte(if (avoidCache) 1 else 0)
        parcel.writeByte(if (animatedStickerPack) 1 else 0)
        parcel.writeTypedList(stickers)
        parcel.writeByte(if (isWhatsAppInstalled) 1 else 0)
    }

    override fun describeContents(): Int = 0

    companion object CREATOR : Parcelable.Creator<StickerPack> {
        const val MAX_STICKERS_PER_PACK = 30

        override fun createFromParcel(parcel: Parcel): StickerPack = StickerPack(parcel)
        override fun newArray(size: Int): Array<StickerPack?> = arrayOfNulls(size)
    }
}
